src/expr/relation/mod.rs

```diff
@@ -842,6 +845,84 @@ impl RelationExpr {
         self.visit1_mut(|e| e.visit_mut_pre(f))
     }
 
+    /// Mutably visit all `[ScalarExpr::Literal]`s in the relation expression.
+    pub fn visit_literals_mut<F>(&mut self, f: &mut F)
+    where
+        F: FnMut(&mut Result<Row, EvalError>, &mut ColumnType),
+    {
+        let mut visit_scalar = |scalar: &mut ScalarExpr| {
+            scalar.visit_mut(&mut |e| {
+                if let ScalarExpr::Literal(v, t) = e {
+                    f(v, t);
+                }
+            });
+        };
+
+        self.visit_mut(&mut |e| match e {
+            RelationExpr::Map { scalars, input: _ }
+            | RelationExpr::Filter {
+                predicates: scalars,
+                input: _,
+            } => {
+                for s in scalars {
+                    visit_scalar(s);
+                }
+            }
+            RelationExpr::FlatMapUnary {
+                expr,
+                input: _,
+                func: _,
+                demand: _,
+            } => visit_scalar(expr),
+            RelationExpr::Join {
+                equivalences: keys,
+                inputs: _,
+                demand: _,
+                implementation: _,
+            }
+            | RelationExpr::ArrangeBy { input: _, keys } => {
+                for key in keys {
+                    for s in key {
+                        visit_scalar(s);
+                    }
+                }
+            }
+            RelationExpr::Reduce {
+                group_key,
+                aggregates,
+                input: _,
+            } => {
+                for s in group_key {
+                    visit_scalar(s);
+                }
+                for agg in aggregates {
+                    visit_scalar(&mut agg.expr);
+                }
+            }
+            RelationExpr::Constant { rows: _, typ: _ }
```

Comment:
I need to read more to check out the uses, but there is a risk that literals can get promoted into constant elements. One transform takes `Map` expressions with literals and turns them in to joins with a constant relation (to allow movement around joins). Maybe fixed without the sneaky literal use, but wanted to signpost.

Reply:
For posterity: this has been fixed by not hacking in a sentinel literal, but using a dedicated `NullaryFunc`.